ztrouy/loncotes-library
# Advanced Endpoints

### Purpose
To allow users to get all available Materials, get all overdue Checkouts, get all Checkouts that have an unpaid late fee, and pay a late fee

### Files Changed
- Added boolean property Paid to `Checkout.cs`, `CheckoutDTO.cs`
- Created initial DTO class for Checkouts that includes a calculated late fee property in new `CheckoutLateFeeDTO.cs` class
- Created initial DTO class for Patrons that includes a calculated Unpaid Balance property in new `PatronWithBalanceDTO.cs` class
- Updated Checkout instances to include value for Paid property in `LoncotesLibraryDbContext.cs`
- Created new Migration `PaidCheckouts`
- Updated endpoints and DTO generators to handle updated DTOs, added GET endpoint that returns all Materials that aren't checkout out, added GET endpoint that returns all overdue Checkouts, added PUT endpoint that sets a Checkout's Paid property to "true", and added a GET endpoint that returns all Checkouts that need to be paid in `Program.cs`

### To Test
Fetch, and run the following commands in order
```
dotnet ef migrations add InitialCreate
```
```
dotnet ef database update
```
```
dotnet watch run
```
Then confirm the following
- GET endpoint `api/patrons/{id}` includes the Patron's total Balance of unpaid or overdue Checkouts
- GET endpoint `api/materials/available` returns all Materials that aren't currently checked out, with their Material Type and Genre
- GET endpoint `api/checkouts/overdue` returns all Checkouts that haven't been returned yet should have been, including each Checkouts' Patron and Material, and the Materials' Material Type and Genre
- GET endpoint `api/checkouts/unpaid` returns all overdue Checkouts or returned Checkouts with an unpaid late fee, and includes each Checkouts' Patron and Material, and the Materials' Material Type and Genre
- PUT endpoint `api/checkout/{id}/return` that sets the Paid property of a returned Checkout with a late fee to "True"
